### PR TITLE
Intel Aler Lake leak checking rule

### PR DESCRIPTION
**Pull request by phretor**
I added a rule to check if a BIOS image has been signed with leaked keys.

### CODE REVIEW

**Comment by matrosov** [comment]
Hi Federico, could you create a new folder and move your rule there: 
`FwHunt/rules/SupplyChain/IntelAlderLakeLeak.yml`

**Comment by phretor** [comment]
Yup! Makes sense.

**Comment by phretor** [comment]
@matrosov done. Have a look.

**Comment by yeggor** [comment]
Hi @phretor. Unfortunately in this form the rule cannot be merged. You use the rule as follows:

```bash
python fwhunt_scan_analyzer.py scan -r ../FwHunt/rules/Threats/IntelAlderLakeLeak.yml rom.bin
```

The `scan` command is only for checking EFI modules, because it will be analyzed by rizin-based analyser before being scanned. If we will use fwhunt as you do, the UefiAnalyzer from fwhunt-scan will try to analyze the whole firmware as a separate EFI module.
In order to analyze the entire firmware, we introduced the `scan-firmware` command. But to use this command, we have to specify the GUIDs of the files we want to analyze:
https://github.com/binarly-io/FwHunt/blob/cb5d688d25e65fd16009bc27674ed49260af97eb/rules/Vulnerabilities/AMI/BRLY-2022-016.yml#L10

Also, if the component you want to detect is in the compressed range, the rule will not detect anything.
If you send a sample on which the rule works for you, we'll be happy to fix it. In the meantime, it is better to use YARA:
```
rule alder_lake_leak {
  strings:
    $s1 = { 30 82 ... }
    $s2 = { 30 82 ... }
    $s3 = { 30 82 ... }
    $s4 = { 30 82 ... }
    $s5 = { 30 82 ... }
    $s6 = { 30 82 ... }
    $s7 = { 30 82 ... }
    $s8 = { 30 82 ... }
  condition:
    any of them
}
```

**Comment by phretor** [comment]
>  In order to analyze the entire firmware, we introduced the scan-firmware command. But to use this command, we have to specify the GUIDs of the files we want to analyze

Is it possible to specify a "catch all" GUIDs to write such generic rules?

> Also, if the component you want to detect is in the compressed range, the rule will not detect anything.

Out of curiosity, does UEFI allow to put certs/sigs in the compressed range?

>  If you send a sample on which the rule works for you, we'll be happy to fix it.

(un)Fortunately, I still haven't found a sample affected by the leak, but I don't have much ROM images at my disposal. So I "synthetized" a test case by appending the public key (already) found in my laptop's ROM image to the rule and verified that there was a match.

**Comment by yeggor** [comment]
СС @flothrone @xorpse

In fact, FwHunt was not intended as an alternative to grep or YARA. On the contrary, it was designed to detect things that are hard to detect with a byte-search (for example, we can specify which module we want to scan or only scan the SMI handler functions).

> Out of curiosity, does UEFI allow to put certs/sigs in the compressed range?

It really depends on what kind of certs/sigs we're talking about. Keys located in the `BoodGuard Key Manifest` and `BootGuard Boot Policy` will not be in the compressed range.

> (un)Fortunately, I still haven't found a sample affected by the leak, but I don't have much ROM images at my disposal. So I "synthetized" a test case by appending the public key (already) found in my laptop's ROM image to the rule and verified that there was a match.

After a closer look, we concluded that you were looking for the wrong byte patterns.
Instead of DER-formatted keys, you need to look for a public key module with a reverse byte order.
Below is the proposed YARA rule:

```
rule alder_lake_leak {
  strings:
    $s1 = { ff 84 9b 32 ff 8a 95 6b 59 49 86 8d 61 91 01 65 1a 35 ae 51 18 2a 8f 55 05 92 a8 2f f1 4e 96 40 3f 35 c2 fa d4 03 c8 f9 13 10 f0 e4 ad cf 74 7c 62 a0 80 5d 40 d8 80 2e 47 40 24 df fd 02 28 89 91 08 6a d8 18 af b8 3a 96 7d be e0 73 a9 4b 20 fa 09 57 51 e6 be 3d 43 78 c9 94 29 f5 af 93 b1 b3 03 a5 88 6b c7 d7 28 f4 51 ef f0 f2 3a 0a f9 81 2e b6 c5 5b 9b 12 75 fa eb d1 6a ce db 9f 52 b0 8a 5c e7 80 2e 09 71 12 6f a6 91 0a ce 7a 70 b1 32 84 e9 a1 2e 3f 4f 95 3d e9 3e c0 b1 94 1a 2b 7e 6f 47 c7 14 e2 d5 cb 48 1a 42 30 c8 b8 03 28 15 18 3a a3 2e 5b 19 97 14 50 63 17 6d da 64 24 85 c2 71 64 9d 6f e2 90 07 60 b5 27 c8 6f 51 6d e7 3f 5c 77 7a 29 aa 54 17 3a 2f 3b 51 d0 71 3c 4f 4c 8c ee b8 b0 bc 38 69 d5 ab a0 74 37 96 d8 f6 50 01 37 1e ea 7b b1 a1 47 2d 67 ce 5d bf }
    $s2 = { 6b ff 67 03 2a 6b 05 bc ea 29 3a 73 2a 96 56 cb 4e 41 a1 cf e5 e7 1d e2 03 02 ed 33 29 3a 76 0e d4 ed 36 9d 0c cf 37 fa fd c3 bb af 21 dd 3f 52 f3 0c 21 29 ae 0d 20 5a 26 90 10 97 c1 fd 67 dd bf 3a 93 5b 79 07 59 1e 99 ff ce 35 1e ef 3b dc da 87 8d af 25 8e 7d fc 7b 8e e6 be a0 18 4c e4 0a a6 43 79 59 61 f1 d8 c4 18 8b da b2 7b 6e 11 54 41 fb a2 f3 1a b1 a4 c9 fe cd fc 04 82 46 b8 3f e7 61 5b 27 23 9e 15 75 f4 ff a8 a4 42 fc dd d4 6c 43 60 c3 5e b8 7f f3 3d 0a a5 69 fd 28 1d 04 34 47 80 12 d2 36 fb 2e 1e e4 d8 84 06 96 2d a1 6e 5a a4 76 7d 98 2b 8e 83 46 d2 7c 3c dd 80 ab 76 f2 53 35 78 2b 32 7c 0e d9 a0 ed fe a4 bc 9a 81 14 f8 57 47 b1 e0 dc f5 a4 fd 37 06 98 a4 60 38 48 4b 2f 01 d1 3a 9a d1 6f 4f d0 07 3d 0f 2a a2 60 5d be 0c 75 8d 60 b5 6e d8 ae 2b d1 b9 }
    $s3 = { dd d5 d1 ef ba 0b 58 6e 93 3c 3d fb f7 aa 84 de ab a6 71 6c 57 74 7c 51 7b 82 43 88 4f 0f a5 dc 57 eb b2 ed e5 0c 1f 3d fe 5b 07 c9 c3 2c 8f 46 3f cb 73 85 66 74 b7 99 6d f5 67 36 66 be 1a 4e 98 90 84 f3 a5 19 b0 3f 1a 4a d2 97 51 66 ff 4c 75 09 7f 30 0b 32 8f d6 1e 87 9a 38 fb f3 41 c1 b3 4f 89 6b 1a 82 df c5 1b b2 85 7d 64 a8 e0 52 62 1e f5 7e d6 a9 e6 d3 93 9f e9 68 82 6d dc 69 e2 a1 2e 29 3a 56 9f d5 04 3c e3 3b b0 92 69 79 fa 24 07 1b b1 74 85 8d 94 1f 50 39 0f db c4 14 fd 46 69 43 4c 76 41 99 78 dd 01 9b 0c 54 96 37 76 41 cc cf 13 67 5a c5 73 db 0e 52 5c 47 d4 c8 75 ed 8a 73 74 f3 25 60 9f 7c 57 1a 95 e9 ea 0d 14 4f ad cf d6 a5 7c 01 3b 9d ae dd d0 6e 15 68 31 fc a8 33 e8 b3 fd 94 1b 28 0d 59 55 89 42 7f 9a 3e 33 1e 9f 47 b1 50 27 be 6c 96 0f 82 fe ca }
    $s4 = { ad 29 5a 8e 38 f7 79 cf b6 e2 69 3f ea b9 12 63 f0 48 34 2a 1e c2 e8 0f 7e 34 61 b3 d9 05 22 81 2c a4 f0 f4 e0 ae 02 b0 74 6c 5b 32 e7 b5 a0 84 95 8a 50 2b 1b 0c 2a cf 95 7d fa 23 56 95 d0 89 5f fa c8 24 df af 9f 17 52 8c ab 40 11 55 51 c7 1f d5 01 cb fc fa c4 cf b9 14 78 5a 87 ed a6 d7 3b 3c ab a2 04 fa 0d f8 ca 62 2f 52 94 e2 c0 2f fc 7a e3 9f d5 a7 38 b6 f2 29 09 ac 54 42 b5 e7 9d 55 6c b4 a2 cc 83 ba 55 ed 99 0c a8 18 0a 26 e0 09 a7 8a cb 0e 5e 23 43 41 6c 1b a2 28 30 05 35 01 71 bc a0 54 91 64 86 41 36 a8 ff 40 00 66 00 aa fd 42 26 4d 3c 61 4f 3b 93 35 ad 04 b5 03 0e 01 b2 9f ad a7 6b 68 7c c3 ae 6c 12 8d 48 15 86 29 64 fb 32 1a 39 94 64 61 70 78 ca 31 f6 5d 09 65 cb 08 dd 58 d9 52 d4 e3 ad 5c f7 e4 33 4f 5d 08 eb a8 85 57 ed f5 fd 6f 44 6c 17 18 a7 ba }
    $s5 = { 13 15 84 20 e2 d1 fb 84 7e 49 a7 99 4d f9 57 ef df cc 97 5b 82 38 aa 21 91 e9 8c 24 62 0d b1 63 b9 d3 4b a8 f1 e9 6b 40 8e c8 17 32 91 32 4c 25 ad e3 8a 3f 2c 77 08 1e 4f 19 55 01 12 fb 5c 90 55 e2 e5 76 42 ec c0 41 1f e7 ec 96 5f 1e ab 8f 9c 24 d6 26 90 a8 4e 3a 51 e5 14 93 1c 86 04 9b 65 74 43 8d 99 7c 79 50 31 36 e9 97 91 92 c6 6b a8 aa 4f 7d b7 7a 76 0f ba 35 21 74 b6 04 8b f5 a0 82 0c 0b 30 87 e3 2f 75 d7 2e f2 7d dc 70 e8 b2 80 7d 39 f1 c4 91 40 ec 12 5e 7a 44 5c c7 22 c7 80 9e 56 3f 99 09 35 ce 33 f8 e4 b2 5c 44 33 fd d7 3d 31 ec 21 36 fd 12 06 f1 96 78 cb ce a8 b6 4c 58 c2 b7 a8 dd ea 5d ae e5 4a 28 9b 5f 6c 8a b8 f8 69 8d 97 7a 24 1c bf 74 c5 b1 02 ea 0e 65 0b d0 d5 40 5f 6b dc 9b cf ec 40 d1 dd 6a f2 c7 29 4e b8 80 a1 6b cf 8e 3b 1c 28 40 69 e0 e4 }
    $s6 = { ff 78 7c 6e c3 d0 d9 ab af 5b 5d 5c b8 7d e3 f4 37 1b e7 2d a1 0a 9d 04 d0 7c 8e 98 5e 71 b2 09 6e d6 6f 9a c9 79 30 17 2a 97 90 6d 6e 73 16 ca 50 02 4f 2c 85 bb 49 f3 5e bf 7b 18 1c 42 0a 48 07 81 10 2a b2 bb 29 38 a2 eb c6 48 27 03 1e a5 eb 18 63 7d 76 43 be 58 58 65 c5 a1 b4 7f a4 ee 67 10 7e 45 3d ad c8 91 b9 b7 fe e8 26 a2 65 79 bc ea ab 64 02 e2 54 99 bb 89 6e 74 52 fa da 98 bf 7d ab 15 86 ad 95 af f4 ab a2 e0 dd 79 95 00 96 38 2a 63 1e 99 a2 10 dc 22 06 73 6a 4d 14 f0 d5 9a 89 8f 07 81 23 4a bb f9 fa 9f f2 01 a4 e4 c1 17 15 cb fe 47 2e 95 26 82 97 0c 39 95 dc ca 1f 59 91 19 c7 a0 5d f8 cb f6 62 74 5a 62 0d ca 35 8d 93 0c 2e 3b 12 6f 1d 8f d3 64 ac d3 c8 51 ff c7 31 94 39 25 99 43 31 07 44 86 ed 36 20 c3 aa a4 3c d4 46 0a ef 2d 1f a3 90 b8 cd cf 31 13 e0 86 7e ad 12 21 88 d2 be b3 63 4b 2c 5c e1 2c 8e cb ea 3b e7 2b f1 d4 50 c1 50 c5 7d 58 4e f5 3b 58 54 69 f1 a2 e8 6a d6 e7 b4 d2 70 c2 5a 36 fd 9b aa b1 f2 bb 3e 1d 6d e2 74 15 09 5b c0 0a 2c 05 90 59 71 eb 6f b5 9f cd fe c6 1e 9d 38 28 da 56 9d cf 05 82 69 5c 34 57 e9 8c 27 3b ae cb 4f 82 26 d3 ee 35 bd 8a 77 ad c8 39 00 8a f3 5c 2f 8e 8e f6 69 54 74 b2 23 50 17 04 0f 78 9c eb }
    $s7 = { d9 03 fc 44 eb ad 15 79 bf b1 a5 45 22 f2 af a8 6e bd b8 62 f5 4b 59 fe 6b 97 a6 9a f0 74 59 89 e2 35 27 00 74 7e 8b 10 df ef 15 84 d0 d9 a7 77 e8 af f3 7e de 4a 2a 1a 18 5f 50 ed 01 b7 4d a4 bd b4 65 a5 74 88 10 8a 22 f6 b0 c6 e6 a1 ba 64 5e d8 5e 8f fc 91 37 ef fa 88 66 56 40 1d 9a 60 43 91 dc 0b 6d 8b 01 28 4a 5b 4d b7 1f fc 0f 79 8e 92 b4 03 0b 02 b8 3b 16 ba d3 a7 f4 70 72 d8 4e e7 c0 0c 52 57 b1 05 74 c7 24 d2 6b c6 b7 5a ba 35 6e 81 0f ca 0c 46 cf 8f bb f4 8d fc 5b 3d 85 59 b0 35 7b 30 c2 10 4e 93 3c 6e cc 66 cc 2d d1 4f 5a 5e ce 73 4c 25 78 f1 73 4c e2 25 33 25 18 9c 63 9b 21 58 1f c5 6f aa 40 36 58 25 78 a4 a8 6d c5 ca 5f 95 11 d2 03 6e 00 fa 74 61 9d 2b 18 16 41 0a 3c cf 84 e8 bb b4 a7 76 d3 d9 86 2b 42 68 f3 1e 31 31 4e ad 28 ec f6 66 53 e5 26 9f c6 fa 39 6d 17 04 bf 5b d3 3e 55 24 0e b1 1f 20 90 60 8d 97 c5 b3 b7 ed dc 9f 46 9f 2f 62 5d 10 e9 80 e8 4c cc 0d 64 e0 1a b2 11 d6 03 44 24 aa 41 14 07 28 0d e3 a8 a6 e7 27 17 23 65 88 46 ed 9b de 9a c2 37 e2 2f 4f 14 3d 32 2a e2 0e 2c 41 36 7a d6 9b f1 a4 ea 8d 26 a9 a8 85 f0 56 69 00 65 88 47 ca 3b 7d 17 f6 7b e9 cb 5a 49 39 8b 41 f0 f6 5b d0 11 30 c8 92 c8 98 51 af c6 4a 76 d8 }
    $s8 = { 53 45 0f f5 91 47 0c 7d c3 d7 43 0b 3f 99 78 0d 3d 6c 87 b5 4b 32 22 4d 6a 1c a8 83 94 c4 c0 83 13 93 d9 d6 bd 5f 9c 2c 2e df 86 95 9c aa a3 69 17 c0 5e f4 2b 9e 0f c1 6c 96 6e 0f df 96 80 14 fe 51 a7 88 63 13 ea 80 48 15 d5 ed 24 9e 7d fe 17 7c 51 8c 02 3b 28 0f cf 62 d7 13 e0 48 95 fc 3b e6 7c 5f de e2 6f d3 f5 02 a6 0c 8b e9 cc ad ef e6 32 d5 38 07 ee 7e 67 87 59 05 23 9b c8 8b 84 c0 2f aa 4e 09 5c 77 8e 50 87 79 4e c6 c3 6f 39 f9 50 c8 93 5c a6 2c 4c 2e 3f 2a c3 02 5a 9c 6d 0d 4e 3e 5b 3c f8 5d c7 8d af c7 ea ad cd 2f 13 37 ff 4f 8c f3 fa e4 8b c7 b6 31 e4 64 a9 23 dd 1f 60 2f 17 a2 22 89 c2 f4 ec 56 57 28 9d 80 18 31 ee c8 4c f8 a7 90 87 e4 7a 7a fb f3 f0 85 ae 42 b0 54 d4 a3 fd 2c 4f 87 82 79 46 ea 4f a2 45 81 b7 cd 76 33 49 6e 5d b8 93 85 a9 47 e2 c0 7b 24 eb 7f d9 bd a2 66 f6 ab e6 07 89 e1 de ee 1a 27 c9 cc 33 cb 86 e0 a6 2a d8 e5 fc f2 a0 cc 52 36 8b c7 3b 2d c2 09 cf 2a 6a d3 80 e5 97 de c8 84 05 1a f8 88 30 42 67 5d 95 61 7a 8b 99 02 44 69 48 e8 96 c1 a1 37 5b 64 fb ce 46 f0 b3 92 7f 77 d1 b1 b6 b3 d6 e0 06 9d f3 07 af c7 e6 d6 0e 11 39 96 03 8c 34 8b bf 0e a8 da b5 02 5a 61 a1 40 34 99 57 e8 a0 05 89 a6 e7 b9 cb 50 43 e3 }
  condition:
    any of them
}
```

If you are still interested in adding a fwhunt rule, we can do it as follows:

```
IntelAlderLakeLeak:
  meta:
    author: Federico Maggi (https://twitter.com/phretor)
    license: CC0-1.0
    name: IntelAlderLakeLeak
    namespace: SupplyChain
    description: BIOS image signed with leaked private keys
    url: https://github.com/phretor/intel-leak-checker/
    target: firmware # so that fwhunt-scan does not try to analyze the entire firmware
  hex_strings:
    or:
      - ff849b32ff8a956b5949868d619101651a35ae51182a8f550592a82ff14e96403f35c2fad403c8f91310f0e4adcf747c62a0805d40d8802e474024dffd02288991086ad818afb83a967dbee073a94b20fa095751e6be3d4378c99429f5af93b1b303a5886bc7d728f451eff0f23a0af9812eb6c55b9b1275faebd16acedb9f52b08a5ce7802e0971126fa6910ace7a70b13284e9a12e3f4f953de93ec0b1941a2b7e6f47c714e2d5cb481a4230c8b8032815183aa32e5b1997145063176dda642485c271649d6fe2900760b527c86f516de73f5c777a29aa54173a2f3b51d0713c4f4c8ceeb8b0bc3869d5aba0743796d8f65001371eea7bb1a1472d67ce5dbf
      - 6bff67032a6b05bcea293a732a9656cb4e41a1cfe5e71de20302ed33293a760ed4ed369d0ccf37fafdc3bbaf21dd3f52f30c2129ae0d205a26901097c1fd67ddbf3a935b7907591e99ffce351eef3bdcda878daf258e7dfc7b8ee6bea0184ce40aa643795961f1d8c4188bdab27b6e115441fba2f31ab1a4c9fecdfc048246b83fe7615b27239e1575f4ffa8a442fcddd46c4360c35eb87ff33d0aa569fd281d0434478012d236fb2e1ee4d88406962da16e5aa4767d982b8e8346d27c3cdd80ab76f25335782b327c0ed9a0edfea4bc9a8114f85747b1e0dcf5a4fd370698a46038484b2f01d13a9ad16f4fd0073d0f2aa2605dbe0c758d60b56ed8ae2bd1b9
      - ddd5d1efba0b586e933c3dfbf7aa84deaba6716c57747c517b8243884f0fa5dc57ebb2ede50c1f3dfe5b07c9c32c8f463fcb73856674b7996df5673666be1a4e989084f3a519b03f1a4ad2975166ff4c75097f300b328fd61e879a38fbf341c1b34f896b1a82dfc51bb2857d64a8e052621ef57ed6a9e6d3939fe968826ddc69e2a12e293a569fd5043ce33bb0926979fa24071bb174858d941f50390fdbc414fd4669434c76419978dd019b0c5496377641cccf13675ac573db0e525c47d4c875ed8a7374f325609f7c571a95e9ea0d144fadcfd6a57c013b9daeddd06e156831fca833e8b3fd941b280d595589427f9a3e331e9f47b15027be6c960f82feca
      - ad295a8e38f779cfb6e2693feab91263f048342a1ec2e80f7e3461b3d90522812ca4f0f4e0ae02b0746c5b32e7b5a084958a502b1b0c2acf957dfa235695d0895ffac824dfaf9f17528cab40115551c71fd501cbfcfac4cfb914785a87eda6d73b3caba204fa0df8ca622f5294e2c02ffc7ae39fd5a738b6f22909ac5442b5e79d556cb4a2cc83ba55ed990ca8180a26e009a78acb0e5e2343416c1ba2283005350171bca0549164864136a8ff40006600aafd42264d3c614f3b9335ad04b5030e01b29fada76b687cc3ae6c128d4815862964fb321a399464617078ca31f65d0965cb08dd58d952d4e3ad5cf7e4334f5d08eba88557edf5fd6f446c1718a7ba
      - 13158420e2d1fb847e49a7994df957efdfcc975b8238aa2191e98c24620db163b9d34ba8f1e96b408ec8173291324c25ade38a3f2c77081e4f19550112fb5c9055e2e57642ecc0411fe7ec965f1eab8f9c24d62690a84e3a51e514931c86049b6574438d997c79503136e9979192c66ba8aa4f7db77a760fba352174b6048bf5a0820c0b3087e32f75d72ef27ddc70e8b2807d39f1c49140ec125e7a445cc722c7809e563f990935ce33f8e4b25c4433fdd73d31ec2136fd1206f19678cbcea8b64c58c2b7a8ddea5daee54a289b5f6c8ab8f8698d977a241cbf74c5b102ea0e650bd0d5405f6bdc9bcfec40d1dd6af2c7294eb880a16bcf8e3b1c284069e0e4
      - ff787c6ec3d0d9abaf5b5d5cb87de3f4371be72da10a9d04d07c8e985e71b2096ed66f9ac97930172a97906d6e7316ca50024f2c85bb49f35ebf7b181c420a480781102ab2bb2938a2ebc64827031ea5eb18637d7643be585865c5a1b47fa4ee67107e453dadc891b9b7fee826a26579bceaab6402e25499bb896e7452fada98bf7dab1586ad95aff4aba2e0dd79950096382a631e99a210dc2206736a4d14f0d59a898f0781234abbf9fa9ff201a4e4c11715cbfe472e952682970c3995dcca1f599119c7a05df8cbf662745a620dca358d930c2e3b126f1d8fd364acd3c851ffc731943925994331074486ed3620c3aaa43cd4460aef2d1fa390b8cdcf3113e0867ead122188d2beb3634b2c5ce12c8ecbea3be72bf1d450c150c57d584ef53b585469f1a2e86ad6e7b4d270c25a36fd9baab1f2bb3e1d6de27415095bc00a2c05905971eb6fb59fcdfec61e9d3828da569dcf0582695c3457e98c273baecb4f8226d3ee35bd8a77adc839008af35c2f8e8ef6695474b2235017040f789ceb
      - d903fc44ebad1579bfb1a54522f2afa86ebdb862f54b59fe6b97a69af0745989e2352700747e8b10dfef1584d0d9a777e8aff37ede4a2a1a185f50ed01b74da4bdb465a57488108a22f6b0c6e6a1ba645ed85e8ffc9137effa886656401d9a604391dc0b6d8b01284a5b4db71ffc0f798e92b4030b02b83b16bad3a7f47072d84ee7c00c5257b10574c724d26bc6b75aba356e810fca0c46cf8fbbf48dfc5b3d8559b0357b30c2104e933c6ecc66cc2dd14f5a5ece734c2578f1734ce2253325189c639b21581fc56faa4036582578a4a86dc5ca5f9511d2036e00fa74619d2b1816410a3ccf84e8bbb4a776d3d9862b4268f31e31314ead28ecf66653e5269fc6fa396d1704bf5bd33e55240eb11f2090608d97c5b3b7eddc9f469f2f625d10e980e84ccc0d64e01ab211d6034424aa411407280de3a8a6e7271723658846ed9bde9ac237e22f4f143d322ae20e2c41367ad69bf1a4ea8d26a9a885f0566900658847ca3b7d17f67be9cb5a49398b41f0f65bd01130c892c89851afc64a76d8
      - 53450ff591470c7dc3d7430b3f99780d3d6c87b54b32224d6a1ca88394c4c0831393d9d6bd5f9c2c2edf86959caaa36917c05ef42b9e0fc16c966e0fdf968014fe51a7886313ea804815d5ed249e7dfe177c518c023b280fcf62d713e04895fc3be67c5fdee26fd3f502a60c8be9ccadefe632d53807ee7e67875905239bc88b84c02faa4e095c778e5087794ec6c36f39f950c8935ca62c4c2e3f2ac3025a9c6d0d4e3e5b3cf85dc78dafc7eaadcd2f1337ff4f8cf3fae48bc7b631e464a923dd1f602f17a22289c2f4ec5657289d801831eec84cf8a79087e47a7afbf3f085ae42b054d4a3fd2c4f87827946ea4fa24581b7cd7633496e5db89385a947e2c07b24eb7fd9bda266f6abe60789e1deee1a27c9cc33cb86e0a62ad8e5fcf2a0cc52368bc73b2dc209cf2a6ad380e597dec884051af8883042675d95617a8b9902446948e896c1a1375b64fbce46f0b3927f77d1b1b6b3d6e0069df307afc7e6d60e113996038c348bbf0ea8dab5025a61a140349957e8a00589a6e7b9cb5043e3
```

**Comment by yeggor** [comment]
After these fixes, the rule above can be merged:
1. https://github.com/binarly-io/fwhunt-scan/commit/e3a036f3cbd576d5667c548e492dd97c245c7fe5
2. https://github.com/binarly-io/fwhunt-scan/commit/5cea7f0f03013cf4849620db7fe7604348175a87

**Comment by phretor** [comment]
> СС @flothrone @xorpse
> 
> In fact, FwHunt was not intended as an alternative to grep or YARA. On the contrary, it was designed to detect things that are hard to detect with a byte-search (for example, we can specify which module we want to scan or only scan the SMI handler functions).

Ou yes :-) I'm aware of FwHunt/efiXplorer's unique scope and my first thought was also to create a `grep`-based checker (*). But once I was done with that I thought: wait, what's the first thing that comes to mind to check if a firmware has issues? I knew that FwHunt was a bit of a stretch for the task, but I think that FwHunt sees more firmware images passing under it than Yara does, globally.

I didn't think of Yara because I'm not sure how much it's used in firmware DFIR tasks, but to be honest I don't have data to back my position.

(*) = thanks for fixing my naïve patterns..eh!

> > Out of curiosity, does UEFI allow to put certs/sigs in the compressed range?
> 
> It really depends on what kind of certs/sigs we're talking about. Keys located in the `BoodGuard Key Manifest` and `BootGuard Boot Policy` will not be in the compressed range.
> 
> > (un)Fortunately, I still haven't found a sample affected by the leak, but I don't have much ROM images at my disposal. So I "synthetized" a test case by appending the public key (already) found in my laptop's ROM image to the rule and verified that there was a match.
> 
> After a closer look, we concluded that you were looking for the wrong byte patterns. Instead of DER-formatted keys, you need to look for a public key module with a reverse byte order. Below is the proposed YARA rule:

Do you happen to have a firmware image (that you can somehow share) that matches these patterns? It'll be very helpful as a learning example to me. I'll see if VT has something in the meanwhile.

> ```
> rule alder_lake_leak {
>   8<-------8<-----------------------------------------
>   condition:
>     any of them
> }
> ```
> 
> If you are still interested in adding a fwhunt rule, we can do it as follows:
> 
> ```
> IntelAlderLakeLeak:
>   meta:
>     author: Federico Maggi (https://twitter.com/phretor)
>     license: CC0-1.0
>     name: IntelAlderLakeLeak
>     namespace: SupplyChain
>     description: BIOS image signed with leaked private keys
>     url: https://github.com/phretor/intel-leak-checker/
>     target: firmware # so that fwhunt-scan does not try to analyze the entire firmware
>   hex_strings:
>     or:
>   8<-------8<-----------------------------------------
> ```

Thanks! I'll check the changes you pushed in FwHunt.

File: rules/SupplyChain/IntelAlderLakeLeak.yml [hunk 1]
```diff
@@ -3,17 +3,17 @@ IntelAlderLakeLeak:
     author: Federico Maggi (https://twitter.com/phretor)
     license: CC0-1.0
     name: IntelAlderLakeLeak
-    namespace: threats
+    namespace: IntelAlderLakeLeak
```
Comment (yeggor) [inline rules/SupplyChain/IntelAlderLakeLeak.yml]: SupplyChain

**Review by yeggor — CHANGES_REQUESTED** [review]

**Review by phretor — COMMENTED** [review]
Renamed

**Review by phretor — COMMENTED** [review]
Done.

**Comment by phretor** [comment]
Done here https://github.com/binarly-io/FwHunt/commit/6c745ad5d04006f931186a0e2bde194c40da1d9c

**Review by yeggor — APPROVED** [review]

**Comment by yeggor** [comment]
> Do you happen to have a firmware image (that you can somehow share) that matches these patterns? It'll be very helpful as a learning example to me. I'll see if VT has something in the meanwhile.

It might be a good idea to look in the same place where you found the keys.

Thanks a lot for your contribution!

**Comment by phretor** [comment]
> It might be a good idea to look in the same place where you found the keys.

Ah, of course. That's the easiest place to look. I was so convinced that that was source code that I didn't even think of checking for binaries 😄 

Thanks to you for your debugging!

**Pull request merged**